### DISCUSSION TITLE
Question: How to use handlers with webhooks?

### DISCUSSION THREAD

**Discussion by danoctua** [python-telegram-bot/python-telegram-bot#2874] [Q&A] [answered]
Hi there. I'm trying to use webhooks with the command handler for a better code structure. I also want to use other features like `CallbackQueryHandler` and others. It looks impossible to do with a webhook.

Do you have any code snippets on how is it possible to use handlers with webhooks? Is it possible?

I want to use webhooks with my Flask web app. I don't want to start a separate thread, but want to keep the bot working by handling requests. But I don't think I'd like to rewrite the whole handler's logic because of this.

Thank you for your response!

**Comment by Bibo-Joshi** [comment, marked as answer]
Hi. Webhooks have little to do with handler - it's just a different way of fetching the updates that are sent to your bot and doesn't really have an effect on how `Dispatcher` (whos responsible for managing the handlers) works. Please have a look at the [wiki page on webhooks](../wiki/Webhooks).

**Reply by danoctua** [reply]
I've gone through this article but didn't find a clear way how to set it up.

When fetching with pulling I'm running the script with `updater.start_pooling()` and every time the updater receives a response with new updates it sends it to the dispatcher which processes data with handlers.

I want to get the same behavior with the `webhooks` - when Telegram reaches out to my endpoint, I want to fetch the body (updates) with the existing dispatcher and created handlers.

Am I right, that this solution should work for this scenario:

```python3
# my_package/bot.py

def setup_bot() -> Bot:
    bot = Bot(token)
    return bot

def setup_dispatcher(bot: Bot) -> Dispatcher:
    dispatcher = Dispatcher(bot, None)
    # set up handlers here
    return dispatcher
 
bot = setup_bot()
dispatcher = setup_dispatcher(bot)

# my_package/routes.py

@app.route("/accept_response_telegram/<token>", methods=["POST"])
def accept_response_telegram_page(token: str):
    update = Update.de_json(request.get_json(force=True), bot)
    dispatcher.process_response(update)
```
Then if I'll set up e.g. `CommandHandler` in the `setup_dispatcher` and will send a command, it will be retrieved as I'll be using the pooling option?

**Reply by danoctua** [reply]
I tried this solution and here's what I got:

```python3
# Declaring the dispatcher

def setup_dispatcher(bot_: Bot) -> Dispatcher:
    dispatcher_ = Dispatcher(bot_, None)
    # Define command handlers here
    dispatcher_.add_handler(CommandHandler('start', TelegramBotService.start_bot))
    dispatcher_.add_handler(CommandHandler('language', TelegramBotService.action_change_language))
    return dispatcher_


bot = Bot(bot_token)
dispatcher = setup_dispatcher(bot)


class TelegramBotService:
    @classmethod
    def start_bot(cls, update: Update, context: CallbackContext) -> None:
        # raising error in the next line
        context.bot.send_message(chat_id=update.message.chat_id, message="Howdy!")
        return cls.action_change_language(update=update, context=context)
```

It looks like the behavior is not the same as while pooling. The first attribute sent to the method is a `Bot` instance, while the second is an `Update`.

**Reply by danoctua** [reply]
From what I see in the code, the context should be initialized automatically, but it isn't.

**Reply by Poolitzer** [reply]
If you get Bot, Update as a parameter, you are using an old version of this library. Double check which you are using e.g. with `telegram.__version__`. ~And for your first snippet, the incoming json body is a list of Updates.~ Nevermind thats just for long polling.

**Reply by danoctua** [reply]
The main reason while I tracerouted it was that I had to force the `use_context=True` parameter even if by default it seems to be in the `Dispatcher` class `__init__` method. Not sure if it's a bug or not, I'll look at it later to understand the logic behind it better.

Thank you for providing a great package, still discovering new ways how to use it. I'll also try to share a code snippet here later in case anyone will look at how to integrate Flask and Telegram bot webhooks because I didn't find a good example for this on the internet. Your documentation is still the best source of information on this package. Appreciate it.

**Reply by kavin-114** [reply]
Hi @danoctua , Did you find any way to use handlers with webhook? I have tried using application.process_update() but got error as `Application was not initialized using Application.initialize()` . So, please guide me if you have resolved the following issues which you have mentioned above. 

Thank you.

**Reply by Bibo-Joshi** [reply]
> Please have a look at the [wiki page on webhooks](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Webhooks).
